src/mlang/backend_ir/bir.ml

```diff
@@ -18,30 +18,72 @@ type rule_id = Mir.rule_id
 
 module RuleMap = Mir.RuleMap
 
-type variable = Mir.Variable.t
+type tgv_id = string
 
-type variable_id = Mir.variable_id
+let default_tgv = "primitif"
 
-module VariableMap = Mir.VariableMap
-module VariableDict = Mir.VariableDict
+type variable = { on_tgv : tgv_id; offset : int; mir_var : Mir.Variable.t }
 
-(* unify SSA variables *)
-let var_from_mir (v : Mir.Variable.t) : variable =
-  match v.origin with Some v -> v | None -> v
+let compare_variable v1 v2 =
+  let c = String.compare v1.on_tgv v2.on_tgv in
+  if c <> 0 then c
+  else
+    let c = Stdlib.compare v1.offset v2.offset in
+    if c <> 0 then c else Mir.Variable.compare v1.mir_var v2.mir_var
+
+module VariableMap = Map.Make (struct
+  type t = variable
+
+  let compare = compare_variable
+end)
+
+module VariableSet = Set.Make (struct
+  type t = variable
+
+  let compare = compare_variable
+end)
 
-let var_to_mir v = v
+module NameMap = Map.Make (String)
+
+type offset_alloc = { mutable name_map : int NameMap.t; mutable size : int }
+
+(* Mutable state hidden away in the signature. Used for black-magicaly
+   transition variable representations from SSA duplications to offsets of TGV.
+   An issue though: this disregards tetantives to reduce the size of the TGV
```

Comment:
I guess we are kind of stuck with keeping the TGV expected by the clients. However we can still get rid of the computations?

Reply:
Eventually ? Optimisations are disable for most backends right now (since #111), but we could rework an inlining/specialisation decision at some point, would the need arise.

Also, the TGV representation in most backend is different from the current DGFIP workflow nonetheless, but it shouldn't matter as long as clients use the provided interface (currently, mostly `desc.c`).